### PR TITLE
Allow serving API through TLS w/ client authentication

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

#1503 

# Overview

- 3e2aeb8f2158733a1825cccec58a029cf2a411f6
  :round_pushpin: **create new module for generate TLS settings for client auth**
  This is mostly copied from cardano-sl and re-arranged a bit. This validation is very
basic, but so is our TLS setup anyway. In a new module because it's quite isolated
while quite verbose.

- dfcf26642e28861db25cf6352c7901f39a44d897
  :round_pushpin: **run server socket through TLS when provided with a TLS configuration**
  Passing 'Nothing' as an argument will preserve the current behavior where the socket run with
no TLS activated. This is still useful for current integration tests and for the Jörmungandr
setup.

- f0afdcf8419e6b9e2f57feac45282f8f58782e03
  :round_pushpin: **add command-line option for parsing TLS configuration**
  Very simple here, we simply go with path and _assume_ that the given PKI is somewhat
well-formed. It's hard to do any sort of validation here but it'll be good to investigate
what sort of runtime errors we might run into if the provided files aren't certificates...

- 29182d5543cfad19c42f2f47dd120ce77e7da16a
  :round_pushpin: **unit test TLS settings with a mock server**
  Certificates generated using cardano-sl-x509. The test setup is fairly simple, a webserver
is started using the very same 'start' method we use for starting the wallet server.
Then, we simply try to ping it in various scenarios, with and without TLS enabled on the
client side.

- c1d8ea17222108eb54dfe0f9053d1569995f865f
  :round_pushpin: **re-generate nix machinery**

# Comments

:warning: Haven't tested any of this _yet_ ! I'll see if we can simply enable this on our integration setup.

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/Api/Server/Tls.hs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    }
+  where
+    tlsSettings =
+        tlsSettingsChain tlsSvCert [tlsCaCert] tlsSvKey
+
+    -- NOTE
+    -- This checks makes sense only for remote services, to validate that the
+    -- fully qualified hostname from the certificate matches the one from the
+    -- service we're trying to reach. This is of little use for a server
+    -- validation.
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Api/Server/Tls.hs:62]: :+1:

File: lib/core/src/Cardano/Wallet/Api/Server/Tls.hs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    serviceID =
+        ("", "")
+
+    certificateUsageFromValidations =
+        maybe CertificateUsageAccept (CertificateUsageReject . CertificateRejectOther)
+
+    -- By default, X509.Validation validates the certificate names against the host
+    -- which is irrelevant when checking the client certificate (but relevant for
+    -- the client when checking the server's certificate).
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Api/Server/Tls.hs:71]: :+1:

**Review by paweljakubas — APPROVED** [review]
looks reasonable, here assumptions why to omit some validations. PR needs some nix patch application. But would be great to add **some testing** for it to gain confidence it works as expected

**Comment by rhyslbw** [comment]
Getting a build error on this commit hash
```
Configuring library for cardano-wallet-core-2020.3.16..
Warning: The flag --disable-executable-profiling is deprecated. Please use
--disable-profiling instead.
Setup: Encountered missing dependencies:
data-default -any,
tls -any,
warp-tls -any,
x509 -any,
x509-store -any,
x509-validation -any

builder for '/nix/store/f56pb8bmf6nv532s4496x6zqph2n1gy6-cardano-wallet-core-2020.3.16-lib-cardano-wallet-core.drv' failed with exit code 1
cannot build derivation '/nix/store/0hvb1p3hk9f2s6h1aa4bjrag0bd1p7lp-cardano-wallet-byron-2020.3.16-exe-cardano-wallet-byron-config.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/6xf4nd9smmb0r9h65fcgdylllzn4va3m-cardano-wallet-jormungandr-2020.3.16-exe-cardano-wallet-jormungandr-config.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/v7ql9g6nks3srhzhj725g4b0fb398yz9-cardano-wallet-byron-2020.3.16-exe-cardano-wallet-byron.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/9hhhjgia3jyysgbjdqg5vpvihgv01zay-cardano-wallet-jormungandr-2020.3.16-exe-cardano-wallet-jormungandr.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/vzhqjspx6gk1gqw8w8xdpk3iy32nr755-cardano-wallet-byron-2020.3.16.drv': 1 dependencies couldn't be built
error: build of '/nix/store/j5m6cnz1yg3b16bga55x20071gzgqnw0-cardano-wallet-jormungandr-2020.3.16.drv', '/nix/store/vzhqjspx6gk1gqw8w8xdpk3iy32nr755-cardano-wallet-byron-2020.3.16.drv' failed
```

**Comment by KtorZ** [comment]
@rhyslbw , yes as pawel pointed out, we need to apply the patch for re-generating the nix files. I've done it locally, let me push.

File: lib/cli/src/Cardano/CLI.hs [hunk 3]
```diff
@@ -1672,6 +1673,28 @@ addressIndexOption = optionT $ mempty
...
 
+tlsOption
+    :: Parser TlsConfiguration
+tlsOption = TlsConfiguration
+    <$> tlsCaCertOption
+    <*> tlsSvCertOption
+    <*> tlsSvKeyOption
+  where
+    tlsCaCertOption = optionT $ mempty
+        <> long "--tls-ca-cert"
```
Comment (rhyslbw) [inline lib/cli/src/Cardano/CLI.hs]:
```suggestion
        <> long "tls-ca-cert"
```

File: lib/cli/src/Cardano/CLI.hs [hunk 4]
```diff
@@ -1672,6 +1673,28 @@ addressIndexOption = optionT $ mempty
...
+        <> metavar "FILE"
+        <> help "A x.509 Certificate Authority (CA) certificate."
+
+    tlsSvCertOption = optionT $ mempty
+        <> long "--tls-sv-cert"
```
Comment (rhyslbw) [inline lib/cli/src/Cardano/CLI.hs]:
```suggestion
        <> long "tls-sv-cert"
```

File: lib/cli/src/Cardano/CLI.hs [hunk 5]
```diff
@@ -1672,6 +1673,28 @@ addressIndexOption = optionT $ mempty
...
+        <> metavar "FILE"
+        <> help "A x.509 Server (SV) certificate."
+
+    tlsSvKeyOption = optionT $ mempty
+        <> long "--tls-sv-key"
```
Comment (rhyslbw) [inline lib/cli/src/Cardano/CLI.hs]:
```suggestion
        <> long "tls-sv-key"
```

**Review by rhyslbw — DISMISSED** [review]
Currently
```
Available options:
  -h,--help                Show this help text
  --help-tracing           Show help for tracing options
  --listen-address HOST    Specification of which host to the bind API server
                           to. Can be an IPv[46] address, hostname, or
                           '*'. (default: 127.0.0.1)
  --random-port            serve wallet API on any available port (conflicts
                           with --port)
  --port INT               port used for serving the wallet API. (default: 8090)
  ----tls-ca-cert FILE     A x.509 Certificate Authority (CA) certificate.
  ----tls-sv-cert FILE     A x.509 Server (SV) certificate.
  ----tls-sv-key FILE      The RSA Server key which signed the x.509 server
                           certificate.
...
```

File: lib/cli/src/Cardano/CLI.hs (hunk 3 above)
Comment (KtorZ) [reply lib/cli/src/Cardano/CLI.hs]: Whola :man_facepalming:

**Comment by KtorZ** [comment]
@paweljakubas added unit test with a mock server running with TLS settings, using the same `Cardano.Wallet.Api.Server#start` method that the wallet server uses to run the application, but using a mock Wai Application. 

```
Cardano.Wallet.Api.Server.Tls
  TLS Client Authentication
    Respond to authenticated client if TLS is enabled
    Deny client with wrong certificate if TLS is enabled
    Properly deny HTTP connection if TLS is enabled
```

File: lib/cli/src/Cardano/CLI.hs (hunk 3 above)
Comment (KtorZ) [reply lib/cli/src/Cardano/CLI.hs]: commit [amended](https://github.com/input-output-hk/cardano-wallet/compare/c1d8ea17222108eb54dfe0f9053d1569995f865f..c6cad2975fa580ff662ea0a440c441e3c5c8a3df)

File: lib/cli/src/Cardano/CLI.hs (hunk 4 above)
Comment (KtorZ) [reply lib/cli/src/Cardano/CLI.hs]: commit [amended](https://github.com/input-output-hk/cardano-wallet/compare/c1d8ea17222108eb54dfe0f9053d1569995f865f..c6cad2975fa580ff662ea0a440c441e3c5c8a3df)

File: lib/cli/src/Cardano/CLI.hs (hunk 5 above)
Comment (KtorZ) [reply lib/cli/src/Cardano/CLI.hs]: commit [amended](https://github.com/input-output-hk/cardano-wallet/compare/c1d8ea17222108eb54dfe0f9053d1569995f865f..c6cad2975fa580ff662ea0a440c441e3c5c8a3df)

**Comment by KtorZ** [comment]
bors try

**Pull request merged**